src/rule_definition_tools.jl

```diff
@@ -1,5 +1,90 @@
 # These are some macros (and supporting functions) to make it easier to define rules.
 
+"""
+    propagator_name(f, propname)
+
+Determines a reasonable name for the propagator function.
+The name doesn't really matter too much as it is a local function to be returned
+by `frule` or `rrule`, but a good name make debugging easier.
+`f` should be some form of AST representation of the actual function,
+`propname` should be either `:pullback` or `:pushforward`
+
+This is able to deal with fairly complex expressions for `f`:
+
+    julia> propagator_name(:bar, :pushforward)
+    :bar_pushforward
+
+    julia> propagator_name(esc(:(Base.Random.foo)), :pullback)
+    :foo_pullback
+"""
+propagator_name(f::Expr, propname::Symbol) = propagator_name(f.args[end], propname)
+propagator_name(fname::Symbol, propname::Symbol) = Symbol(fname, :_, propname)
+propagator_name(fname::QuoteNode, propname::Symbol) = propagator_name(fname.value, propname)
+
+
+"""
+    propagation_expr(𝒟, Δs, ∂s)
+
+    Returns the expression for the propagation of
+    the input gradient `Δs` though the partials `∂s`.
+
+    𝒟 is an expression that when evaluated returns the type-of the input domain.
+    For example if the derivative is being taken at the point `1` it returns `Int`.
+    if it is taken at `1+1im` it returns `Complex{Int}`.
+    At present it is ignored for non-Wirtinger derivatives.
+"""
+function propagation_expr(𝒟, Δs, ∂s)
+    wirtinger_indices = findall(∂s) do ex
+        Meta.isexpr(ex, :call) && ex.args[1] === :Wirtinger
+    end
+    ∂s = map(esc, ∂s)
+    if isempty(wirtinger_indices)
+        return standard_propagation_expr(Δs, ∂s)
+    else
+        return wirtinger_propagation_expr(𝒟, wirtinger_indices, Δs, ∂s)
+    end
+end
+
+function standard_propagation_expr(Δs, ∂s)
+    # This is basically Δs ⋅ ∂s
+
+    # Notice: the thunking of `∂s[i] (potentially) saves us some computation
+    # if `Δs[i]` is a `AbstractDifferential` otherwise it is computed as soon
+    # as the pullback is evaluated
+    ∂_mul_Δs = [:(@thunk($(∂s[i])) * $(Δs[i])) for i in 1:length(∂s)]
+    return :(+($(∂_mul_Δs...)))
+end
+
+function wirtinger_propagation_expr(𝒟, wirtinger_indices, Δs, ∂s)
+    ∂_mul_Δs_primal = Any[]
+    ∂_mul_Δs_conjugate = Any[]
+    ∂_wirtinger_defs = Any[]
+    for i in 1:length(∂s)
+        if i in wirtinger_indices
+            Δi = Δs[i]
+            ∂i = Symbol(string(:∂, i))
+            push!(∂_wirtinger_defs, :($∂i = $(∂s[i])))
+            ∂f∂i_mul_Δ = :(wirtinger_primal($∂i) * wirtinger_primal($Δi))
+            ∂f∂ī_mul_Δ̄ = :(conj(wirtinger_conjugate($∂i)) * wirtinger_conjugate($Δi))
+            ∂f̄∂i_mul_Δ = :(wirtinger_conjugate($∂i) * wirtinger_primal($Δi))
+            ∂f̄∂ī_mul_Δ̄ = :(conj(wirtinger_primal($∂i)) * wirtinger_conjugate($Δi))
+            push!(∂_mul_Δs_primal, :($∂f∂i_mul_Δ + $∂f∂ī_mul_Δ̄))
+            push!(∂_mul_Δs_conjugate, :($∂f̄∂i_mul_Δ + $∂f̄∂ī_mul_Δ̄))
+        else
```

Comment:
You append to `∂_mul_Δs_primal` and `∂_mul_Δs_conjugate`, but nothing gets entered into `∂_wirtinger_defs`.

Below in the `return` statement we splat `∂_wirtinger_defs`, how will the above (missing values in `∂_wirtinger_defs`) be handled in this return statement? Will the corresponding value for `∂_wirtinger_defs` be `null`? Does it matter?

Reply:
I am not sure, this logic has not been touched in this PR,
and I have never read it closely.
I just moved it around into a new function

Reply:
I think to will not get added to output at all.

I think it can just be deleted in any case

Reply:
Actually I see it is used in the other side of the `if`, so this is needed and it is fine to have it as an empty array, it will just not turn up in the returned expression